### PR TITLE
rewrite runner

### PR DESCRIPTION
**Pull request by uriva**

### CODE REVIEW

File: computation_graph/graph.py [hunk 1]
```diff
@@ -8,6 +8,8 @@
 
 from computation_graph import base_types
 
+remove_future_edges = gamla.compose(tuple, opt_gamla.remove(base_types.edge_is_future))
```
Comment (uriva) [inline computation_graph/graph.py:11]: i think this already exists in line 165

File: computation_graph/graph.py [hunk 2]
```diff
@@ -174,3 +186,28 @@ def source():
...
+        signature=_infer_callable_signature(func),
+        is_terminal=True,
+    )
+
+
+def _aggregator_for_terminal(*args):
+    return tuple(args)
+
+
+DEFAULT_TERMINAL = make_terminal("DEFAULT_TERMINAL", _aggregator_for_terminal)
```
Comment (uriva) [inline computation_graph/graph.py:205]: i think we can remove `DEFAULT_TERMINAL` and `connect_default_terminal` from the api, they are easy to implement outside

File: computation_graph/graph_test.py [hunk 3]
```diff
@@ -366,31 +366,35 @@ def merge(args):
...
-    def node1():
-        return "node1"
-
+    """graph = node1 --> node2 --> DEFAULT_TERMINAL, node1 --> TERMINAL2"""
+    edges = graph.connect_default_terminal(composers.make_compose(_node2, _node1))
     terminal2 = graph.make_terminal("TERMINAL2", gamla.wrap_tuple)
-    g = base_types.merge_graphs(
-        composers.make_compose(lambda x: f"node2({x})", node1),
-        composers.compose_unary(source=node1, destination=terminal2),
+    edges += (graph.make_standard_edge(source=_node1, destination=terminal2),)
```
Comment (uriva) [inline computation_graph/graph_test.py:372]: use `base_types.merge_graphs`

**Review by uriva — COMMENTED** [review]
LGTM

File: computation_graph/composers/__init__.py [hunk 4]
```diff
@@ -193,6 +167,30 @@ def last(*args) -> base_types.GraphType:
...
+    source: base_types.ComputationNode,
+    key: Optional[str],
+    priority: int,
+    is_future: bool,
+    destination: base_types.ComputationNode,
+    unbound_destination_signature: base_types.NodeSignature,
+):
+    if key is None and signature.is_unary(unbound_destination_signature):
+        key = gamla.head(signature.parameters(unbound_destination_signature))
+    assert key is not None and key in signature.parameters(
```
[the remark this answers is not in the document]
Comment (molaxx) [reply computation_graph/composers/__init__.py:181]: if signature is not unary and key was None

[the remark this answers is not in the document]
Comment (uriva) [reply computation_graph/composers/__init__.py:181]: ah right oops

**Pull request merged**